### PR TITLE
Feature/initial implementation

### PR DESCRIPTION
**Pull request by CarlHembrough**
Initial implementation of a dataset API client. Covers only the endpoints currently required by Zebedee.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -8,5 +8,73 @@
...
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <version>5.1.0</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>3.9.0</version>
```
Comment (Matt-Guest) [inline pom.xml]: There now version 3.9.1 from 21/2/2018

File: pom.xml [hunk 2]
```diff
@@ -8,5 +8,73 @@
...
+        </dependency>
+
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.3</version>
```
Comment (Matt-Guest) [inline pom.xml]: Can this be pushed up to 3.7?

File: src/main/java/dp/api/dataset/DatasetAPIClient.java [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * HTTP client for the dataset API.
+ */
+public class DatasetAPIClient implements DatasetClient {
```
Comment (Matt-Guest) [inline src/main/java/dp/api/dataset/DatasetAPIClient.java:38]: Can this class extend closable as well? as CloseableHTTPClient is a internal resource.

File: src/main/java/dp/api/dataset/DatasetAPIClient.java [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+    }
+
+    private String formatErrResponse(HttpRequestBase httpRequest, CloseableHttpResponse response) {
+
+        return String.format("the dataset api returned a %s response for %s",
+                response.getStatusLine().getStatusCode(),
+                httpRequest.getURI());
+    }
+
+    public static void main(String[] args) throws Exception {
```
Comment (Matt-Guest) [inline src/main/java/dp/api/dataset/DatasetAPIClient.java]: Can the main method be moved into a Example class? As anything that imports the client will also take this main.

File: src/main/java/dp/api/dataset/DatasetAPIClient.java [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+        String url = "http://localhost:22000";
+
+        DatasetAPIClient datasetAPIClient = new DatasetAPIClient(url, key);
+
+        Dataset dataset = new Dataset();
+        dataset.setId("123");
+        dataset.setTitle("123");
+
+        //System.out.println(json.writeValueAsString(dataset));
+        //datasetAPIClient.createDataset("123", dataset);
```
Comment (Matt-Guest) [inline src/main/java/dp/api/dataset/DatasetAPIClient.java]: Commented out code here

File: src/main/java/dp/api/dataset/model/Dataset.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+    public String getUri() {
+        return uri;
+    }
+
+    public void setUri(String uri) {
+        this.uri = uri;
+    }
+
+    public static void main(String[] args) throws IOException {
```
Comment (Matt-Guest) [inline src/main/java/dp/api/dataset/model/Dataset.java]: Is this needed?

File: src/main/java/dp/api/dataset/DatasetAPIClient.java [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+
+/**
+ * HTTP client for the dataset API.
+ */
+public class DatasetAPIClient implements DatasetClient {
+
+    private final URI datasetAPIURL;
+    private final String datasetAPIAuthToken;
+
+    private static final String authTokenHeaderName = "internal-token";
```
Comment (Matt-Guest) [inline src/main/java/dp/api/dataset/DatasetAPIClient.java]: This should be "Internal-Token"

File: src/test/java/dp/api/dataset/DatasetAPIClientTest.java [hunk 8]
```diff
@@ -0,0 +1,647 @@
...
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class DatasetAPIClientTest {
+
+    private static final ObjectMapper json = new ObjectMapper();
+
+    private static final String authTokenHeaderName = "internal-token";
```
Comment (Matt-Guest) [inline src/test/java/dp/api/dataset/DatasetAPIClientTest.java]: This should be "Internal-Token"

File: pom.xml (hunk 1 above)
Comment (CarlHembrough) [reply pom.xml]: done

File: pom.xml (hunk 2 above)
Comment (CarlHembrough) [reply pom.xml]: done

File: src/main/java/dp/api/dataset/DatasetAPIClient.java (hunk 3 above)
Comment (CarlHembrough) [reply src/main/java/dp/api/dataset/DatasetAPIClient.java:38]: done

File: src/main/java/dp/api/dataset/DatasetAPIClient.java (hunk 7 above)
Comment (CarlHembrough) [reply src/main/java/dp/api/dataset/DatasetAPIClient.java]: Fixed

File: src/main/java/dp/api/dataset/DatasetAPIClient.java (hunk 4 above)
Comment (CarlHembrough) [reply src/main/java/dp/api/dataset/DatasetAPIClient.java]: Removed main method

File: src/main/java/dp/api/dataset/DatasetAPIClient.java (hunk 5 above)
Comment (CarlHembrough) [reply src/main/java/dp/api/dataset/DatasetAPIClient.java]: Removed main method

File: src/main/java/dp/api/dataset/model/Dataset.java (hunk 6 above)
Comment (CarlHembrough) [reply src/main/java/dp/api/dataset/model/Dataset.java]: No I have removed it 👍

File: src/test/java/dp/api/dataset/DatasetAPIClientTest.java (hunk 8 above)
Comment (CarlHembrough) [reply src/test/java/dp/api/dataset/DatasetAPIClientTest.java]: Fixed

**Review by Matt-Guest — APPROVED** [review]

**Pull request merged**